File: src/api/transcripts.ts
import { callCommand } from "./tauri";
import type { NotebookEntry, TranscriptJob, TranscriptSegment } from "./types";
import type { ListVideoTranscriptJobsInput } from "./generated/ListVideoTranscriptJobsInput";
import type { ResolveTranscriptJobCompanyInput } from "./generated/ResolveTranscriptJobCompanyInput";
import type { UpdateVideoTranscriptJobInput } from "./generated/UpdateVideoTranscriptJobInput";
import type { CreateNoteFromTranscriptSelectionInput } from "./generated/CreateNoteFromTranscriptSelectionInput";
import type { CreateVideoTranscriptJobInput } from "./generated/CreateVideoTranscriptJobInput";
import type { RunVideoTranscriptJobInput } from "./generated/RunVideoTranscriptJobInput";

// Input types GENERATED via ts-rs (ADR 0048); `noteDraft` uses the generated
// TranscriptNoteDraft (shape-identical to the previous inline object).
export type { ListVideoTranscriptJobsInput } from "./generated/ListVideoTranscriptJobsInput";
export type { ResolveTranscriptJobCompanyInput } from "./generated/ResolveTranscriptJobCompanyInput";
export type { UpdateVideoTranscriptJobInput } from "./generated/UpdateVideoTranscriptJobInput";
export type { CreateNoteFromTranscriptSelectionInput } from "./generated/CreateNoteFromTranscriptSelectionInput";
export type { CreateVideoTranscriptJobInput } from "./generated/CreateVideoTranscriptJobInput";
export type { RunVideoTranscriptJobInput } from "./generated/RunVideoTranscriptJobInput";

export function listVideoTranscriptJobs(input: ListVideoTranscriptJobsInput) {
  return callCommand<TranscriptJob[]>("list_video_transcript_jobs", { input });
}

export function listTranscriptSegments(transcriptJobId: string) {
  return callCommand<TranscriptSegment[]>("list_transcript_segments", { transcriptJobId });
}

export function resolveTranscriptJobCompany(input: ResolveTranscriptJobCompanyInput) {
  return callCommand<TranscriptJob>("resolve_transcript_job_company", { input });
}

export function deleteVideoTranscriptJob(jobId: string) {
  return callCommand<void>("delete_video_transcript_job", { jobId });
}

export function updateVideoTranscriptJob(input: UpdateVideoTranscriptJobInput) {
  return callCommand<TranscriptJob>("update_video_transcript_job", { input });
}

export function createNoteFromTranscriptSelection(input: CreateNoteFromTranscriptSelectionInput) {
  return callCommand<NotebookEntry>("create_note_from_transcript_selection", { input });
}

export function createVideoTranscriptJob(input: CreateVideoTranscriptJobInput) {
  return callCommand<TranscriptJob>("create_video_transcript_job", { input });
}

export function runVideoTranscriptJob(input: RunVideoTranscriptJobInput) {
  return callCommand<TranscriptJob>("run_video_transcript_job", { input });
}

File: src-tauri/src/commands/transcripts.rs
use serde::Deserialize;

use crate::{app_state, jobs, storage};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunVideoTranscriptJobInput {
    job_id: String,
    provider_mode: Option<String>,
}

#[tauri::command]
pub fn list_video_transcript_jobs(
    input: storage::TranscriptJobListInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<Vec<storage::TranscriptJob>, String> {
    state
        .list_transcript_jobs(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn delete_video_transcript_job(
    job_id: String,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<(), String> {
    state
        .delete_transcript_job(&job_id)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn create_video_transcript_job(
    input: storage::NewTranscriptJob,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::TranscriptJob, String> {
    state
        .create_transcript_job(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn update_video_transcript_job(
    input: storage::UpdateTranscriptJobInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::TranscriptJob, String> {
    state
        .update_transcript_job(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn list_transcript_segments(
    transcript_job_id: String,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<Vec<storage::TranscriptSegment>, String> {
    state
        .list_transcript_segments(&transcript_job_id)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub fn resolve_transcript_job_company(
    input: storage::ResolveTranscriptJobCompanyInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::TranscriptJob, String> {
    state
        .resolve_transcript_job_company(input)
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn run_video_transcript_job(
    input: RunVideoTranscriptJobInput,
    state: tauri::State<'_, app_state::AppState>,
) -> Result<storage::TranscriptJob, String> {
    let state = state.inner().clone();

    jobs::scheduler::run_blocking_task(move || {
        jobs::transcript_runner::run_video_transcript_job(
            &state,
            &input.job_id,
            input.provider_mode.as_deref(),
        )
    })
    .await
}
